backends/cuda/ceed-cuda.c

```diff
@@ -71,6 +71,8 @@ int CeedCompileCuda(Ceed ceed, const char *source, CUmodule *module,
   char buff[optslen];
   snprintf(buff, optslen,"-arch=compute_%d%d", prop.major, prop.minor);
   opts[numopts + 3] = buff;
+  opts[numopts + 4] = "--maxrregcount=63";
```

Comment:
This doesn't seem to actually be used. Analyzing the output says the kernel used 126 registers. I haven't checked yet if it's coincidence that this is twice the requested number.

Reply:
Interestingly, I commented out the inline option due to the error I mentioned below, and I'm seeing 104 registers/thread. This is also with an older CUDA, so perhaps some changes to nvrtc... The QFunction kernel is the only one above 63, though I tried the same test without the maxrregcount flag, and the QFunction kernel was still the only one above 63.  So it does indeed look like it's not being respected.

I wonder if somehow nvrtc gets confused about usage of the device functions inside the QFunction kernel?  The QFunction inputs and outputs at a quadrature point are read into register arrays through device functions prior to calling the user function. Does nvrtc count this use together with whatever is required within the main user-defined device function?

Reply:
Also, is there a specific reason you chose 63?

Reply:
It was arbitrary, just seeing if it would be used.

Reply:
It's used as a recommendation.

Reply:
I wouldn't recommend using this option, it sometimes result in worst performance.

Reply:
I was thinking we should cap it at *some* size to ensure that blocks could be scheduled at some block size. It doesn't matter here (with shared) because the block sizes can be as small as we choose, but cuda-gen needs specific block sizes (which might need to be bigger than 32).

Also, I see that the cuda-shared `grad` kernel (for BP3, degree 3, Q=5 quadrature points per dimension) might have a block size of `5,5,1` while the corresponding cuda-gen has a block size of `5,5,4`. Probably both would be better off as `5,5,5` (size 125 is close to 128) unless I'm missing something. Anyway, my thought was to be able to find a small viable block size (`5,5,5`) and insist on using few enough registers for it. But this option doesn't seem to be processed as of cuda-11.4 so there's no point pursuing. (We should report it though.)